api/src/Entity/ContentNode.php

```diff
@@ -262,4 +262,36 @@ public function removeChild(self $child): self {
 
         return $this;
     }
+
+    /**
+     * @param ContentNode $prototype
+     */
+    public function copyFromPrototype($prototype) {
+        // copy ContentNode base properties
+        if (!isset($this->contentType)) {
+            $this->contentType = $prototype->contentType;
+        }
+        if (!isset($this->instanceName)) {
+            $this->instanceName = $prototype->instanceName;
+        }
+        if (!isset($this->slot)) {
+            $this->slot = $prototype->slot;
+        }
+        if (!isset($this->position)) {
+            $this->position = $prototype->position;
+        }
+
+        // deep copy children
+        foreach ($prototype->getChildren() as $childPrototype) {
+            $childClass = $childPrototype::class;
+
+            // @var ContentNode $childContentNode
+            $childContentNode = new $childClass();
```

Comment:
Bin mir bewusst, dass das Erstellen von neuen Entitäten etwas im Edge-bereich von dem liegt, was innerhalb einer Doctrine Entity noch akzeptabel ist. Ist aber halt elegant hier, da ich dann direkt `$childContentNode->copyFromPrototype` rekursiv aufrufen kann.

Geht nicht straightforward so elegant, ohne eine Art Entity Service einzuführen. Oder hat jemand eine gute Alternatividee?

Eventuell im Data Persister als separate Funktion, aber dann muss ich mir jeweils den richtigen DataPersister zur Entität dazu suchen.

Reply:
Mir fällt noch das visitor pattern ein, aber das funktioniert in PHP nur so halbpatzig, da kein Method Argument Overloading möglich ist.

Ein leicht anderer Ansatz wäre, nicht `copyFromPrototype` zu programmieren, sondern `clone` oder so etwas ähnliches. Der Vorteil wäre, dass jeder ContentNode nur noch wissen muss, wie er selber erstellt wird, und dann das Erstellen der Children an die Prototyp-Children delegieren kann (ungetesteter Pseudocode):
```php
public function clone() {
  $clone = new self::class();
  $clone->setChildren(array_map($this->getChildren(), fn(ContentNode $child) => $child->clone());

  $clone->contentType = $this->contentType;
  // Etc. für die anderen bestehenden Properties vom Prototype

  return $clone;
}
// DataPersister Code:
$rootContentNode = $data->category->rootContentNode->clone();
$rootContentNode->contentType = $data->contentType ?? $rootContentNode->contentType;
// Etc. für die anderen Properties aus dem Request, falls vorhanden
```
im Vergleich zum entsprechenden Pseudocode deiner bestehenden Variante:
```php
public function copyFromPrototype(ContentNode $prototype) {
  $this->contentType = $this->contentType ?? $prototype->contentType;
  // Etc. für die anderen bestehenden Properties vom Prototype

  $this->setChildren(array_map($prototype->getChildren(), function (ContentNode $prototypeChild) {
    $child = new ($prototypeChild::class);
    $child->copyFromPrototype($prototypeChild);
    return $child;
  }
}
// DataPersister Code:
$rootContentNode = new ColumnLayout();
$rootContentNode->contentType = $data->contentType;
// Etc. für die anderen Properties aus dem Request, falls vorhanden
$rootContentNode->copyFromPrototype($data->category->rootContentNode);
```

Reply:
Hm wenn ich erneut darüber nachdenke, dann ist der grosse Nachteil an der `clone` Variante, dass man z.B. im Fall POST /column_layouts die `$data` vom Request noch manuell auf die geklonte Entity übernehmen muss. Bei der Variante `copyFromPrototype` füllt uns API Platform ja schon schön ein ContentNode mit den Daten vom Request aus.

In beiden Varianten müssen wir bei der Validierung immer berücksichtigen, dass da nach dem Validierungs-Schritt noch die Daten vom Prototype dazukommen werden (dafür haben wir ja die Validierung teilweise schon auf `getXYFromThisOrPrototype` Getter verschoben). Wenn wir irgendwann Validierungen haben, die noch mehrere Properties gleichzeitig miteinbeziehen, dann kann das dann schon mal recht komplex werden.

Die Frage stellt sich, ob wir es überhaupt unterstützen müssen, von einem Prototype zu klonen aber gleichzeitig einzelne Properties manuell zu überschreiben.

Falls wir das unbedingt unterstützen wollen: Eine nochmals ganz andere Variante wäre es vielleicht, es zuzulassen dass man im POST Payload deeply nested Daten mitsendet (`writableLink: true`), und das Kopieren vom Prototype in einem Denormalizer zu machen. Dann wären bei der Validierung die Daten von Payload und Prototype bereits fertig kombiniert.

Reply:
Ich habe in #2218 entdeckt dass es momentan noch nicht gut von Symfony unterstützt wird, wenn man Werte nicht im Payload mitschickt und nur vom Prototype übernehmen will, wegen AutoMapping (automatische Validierung basierend auf Datenbankspalten-Eigenschaften) welches nicht richtig mit Vererbung zusammenarbeitet... Das wäre ein Argument für die Denormalizer + `writableLink: true` Variante...

Reply:
Hmm, guter Punkt.

Bei den meisten Daten dürfte es im Moment eher unkritisch sein, weil die Daten im Prototype ja bereits validiert sind bzw. in der DB konsistent drin stehen.

Kritisch wird es, wenn die properties Abhängigkeiten haben. Beispiel: Auf `ColumnLayout->getColumnsFromThisOrPrototype` haben wir die Validierung `#[AssertNoOrphanChildren]`. Theoretisch könnte ich jetzt von einem Prototype klonen und gleichzeitig das `columns` Property überschreiben. Die Validierung geht durch (weil es zu dem Zeitpunkt noch gar keine Kinder gibt), aber später im data persister füge ich die Kinder dazu - potentiell auf Slots die es nun gar nicht mehr gibt.

**Mögliche Ansätze:**
1. kein gleichzeitiges Überschreiben von properties. Wenn ich von prototype kopiere wird zuerst mal alles überschrieben (ausser natürlich Daten die nicht von `copyFromPrototype` gesetzt werden, die muss ich trotzdem mitliefern). Allfälliges Überschreiben muss ich in einem zweiten Request ausführen. Für Validierung brauche ich stellenweise vielleicht doch `getXYFromThisOrPrototype` (z.B. für Properties die nicht nullable sind).

2.  deeply nested Daten: Wär für mich eine Option für ContentNode. An anderen Stellen (z.B. create camp from prototype) würde das dann aber sehr unpraktisch. 

Von mir aus auch ok, wenn wir uns entscheiden, create from prototype für ContentNode in einem ersten Schritt nicht zu erlauben. Drängt sich für mich nicht zwingend auf für MVP. Was wir auf jeden Fall benötigen ist:
- Create camp from prototype
- Create activity from category

Alles weitere was dann copy & paste innerhalb eines camps (oder auch camp-übergreifend?) erlaubt wär für mich zweite Prio:
- create activity from prototype
- create contentnode from prototype
- ...

Reply:
Ich verstehe deine letzte Überlegung nicht ganz... Für Create Activity from Category brauchen wir doch zwingend intern einen Mechanismus für Create ContentNode from Prototype, nicht?
Nur das überschreiben einzelner Properties während dem Kopieren ist dann kein Thema.

Reply:
Ja, genau. Die Businesslogik zum Kopieren braucht es dennoch. Aber diese wird implizit angestossen (über category property) und nicht explizit (über prototype category).

Eine zusätzliche Validation der kopierten Daten erübrigt sich, da wir ja davon ausgehen, dass die ContentNodes auf der Category ok und validiert sind. Right?

Ähnliches gilt vermutlich auch beim Kopieren eines camps. Da müsste man sicher nochmals im Detail diskutieren, welche Entities dann kopiert werden und welche nicht.